### ISSUE TITLE
Lowest dependencies resolution fail with no vendor and .lock

### ISSUE THREAD

**Issue by Soullivaneuh** [composer/composer#5355]
With the following composer.json:

``` json
{
    "name": "sonata-project/doctrine-orm-admin-bundle",
    "type": "symfony-bundle",
    "description": "Symfony Sonata / Integrate Doctrine ORM into the SonataAdminBundle",
    "keywords": ["Admin Generator", "admin", "sonata", "bootstrap", "generator"],
    "homepage": "https://sonata-project.org/bundles/admin",
    "license": "MIT",
    "authors": [
        {
            "name": "Thomas Rabaix",
            "email": "thomas.rabaix@sonata-project.org",
            "homepage": "https://sonata-project.org"
        },
        {
            "name": "Sonata Community",
            "homepage": "https://github.com/sonata-project/SonataDoctrineORMAdminBundle/contributors"
        }
    ],
    "require": {
        "php": "^5.3 || ^7.0",
        "doctrine/orm": "^2.3",
        "symfony/form": "^2.3 || ^3.0",
        "symfony/security": "^2.3 || ^3.0",
        "symfony/console": "^2.3 || ^3.0",
        "symfony/doctrine-bridge": "^2.2 || ^3.0",
        "symfony/framework-bundle": "^2.2 || ^3.0",
        "symfony/security-acl": "^2.2 || ^3.0",
        "sonata-project/exporter": "^1.3.1",
        "sonata-project/admin-bundle": "^3.0",
        "sonata-project/core-bundle": "^3.0"
    },
    "require-dev": {
        "simplethings/entity-audit-bundle": "~0.1",
        "symfony/phpunit-bridge": "^2.7 || ^3.0"
    },
    "suggest": {
        "simplethings/entity-audit-bundle": "If you want to support for versioning of entities and their associations."
    },
    "provide": {
        "sonata-project/admin-bundle-persistency-layer": "1.0.0"
    },
    "autoload": {
        "psr-4": { "Sonata\\DoctrineORMAdminBundle\\": "" }
    },
    "extra": {
        "branch-alias": {
            "dev-master": "3.x-dev"
        }
    }
}
```

When I run this command:

```
# The issue happen only if composer.lock and vendor are not present
rm -rf vendor composer.lock && composer update --prefer-lowest -vvv
```

I get this output:

```
Reading ./composer.json
Loading config file /home/sullivan/.composer/config.json
Loading config file /home/sullivan/.composer/auth.json
Loading config file ./composer.json
Checked CA file /etc/ssl/certs/ca-certificates.crt: valid
Executing command (/home/sullivan/projects/fork/sonata-project/SonataDoctrineORMAdminBundle): git branch --no-color --no-abbrev -v
Reading /home/sullivan/.composer/composer.json
Loading config file /home/sullivan/.composer/config.json
Loading config file /home/sullivan/.composer/auth.json
Loading config file /home/sullivan/.composer/composer.json
Loading config file /home/sullivan/linux-config/home/.composer/auth.json
Reading /home/sullivan/.composer/vendor/composer/installed.json
Loading plugin Pyrech\ComposerNotifier\NotifierPlugin
Loading plugin Pyrech\ComposerChangelogs\ChangelogsPlugin
Loading plugin SLLH\ComposerVersionsCheck\VersionsCheckPlugin
Running 1.1.0 (2016-05-10 15:21:19) with PHP 5.6.4-4ubuntu6.4 on Linux / 3.19.0-49-generic
Loading composer repositories with package information
Downloading https://packagist.org/packages.json
Writing /home/sullivan/.composer/cache/repo/https---packagist.org/packages.json into cache
Updating dependencies (including require-dev)
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/p-provider-2013.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/p-provider-2014.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/p-provider-2015.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/p-provider-2015-07.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/p-provider-2015-10.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/p-provider-2016-01.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/p-provider-2016-04.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/p-provider-archived.json from cache
Downloading http://packagist.org/p/provider-latest%2476511dd80b185ba9f7d3a4c57d438cdfbd88989a0ae89505af9222cc0aa22400.json
Writing /home/sullivan/.composer/cache/repo/https---packagist.org/p-provider-latest.json into cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-sonata-project$doctrine-orm-admin-bundle.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-doctrine$orm.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-doctrine$dbal.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$console.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-doctrine$common.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$polyfill-mbstring.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-doctrine$inflector.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-doctrine$cache.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-doctrine$collections.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-doctrine$lexer.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-doctrine$annotations.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-doctrine$instantiator.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$form.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$event-dispatcher.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$intl.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$options-resolver.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$property-access.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$icu.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$polyfill-intl-icu.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$polyfill-php54.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$security.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$http-foundation.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$http-kernel.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$polyfill-php55.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-psr$log.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$debug.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-ircmaxell$password-compat.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-paragonie$random-compat.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$security-acl.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$polyfill-php56.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$polyfill-php70.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$polyfill-util.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$security-core.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$doctrine-bridge.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$framework-bundle.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$config.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$filesystem.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$routing.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$stopwatch.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$translation.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$dependency-injection.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$templating.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$finder.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$security-csrf.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$asset.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$class-loader.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$polyfill-apcu.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-sonata-project$exporter.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-sonata-project$admin-bundle.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$validator.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$security-bundle.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$twig-bridge.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$expression-language.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-twig$twig.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-twig$extensions.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-sonata-project$block-bundle.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-sonata-project$core-bundle.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-knplabs$knp-menu-bundle.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-sonata-project$cache.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-cocur$slugify.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-knplabs$knp-menu.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-simplethings$entity-audit-bundle.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$phpunit-bridge.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$security-http.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-symfony$security-guard.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-jms$di-extra-bundle.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-gedmo$doctrine-extensions.json from cache
Reading /home/sullivan/.composer/cache/repo/https---packagist.org/provider-doctrine$doctrine-bundle.json from cache
Resolving dependencies through SAT
Dependency resolution completed in 0.089 seconds
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - sonata-project/admin-bundle 3.0.0 requires knplabs/knp-menu-bundle ^2.1.1 -> satisfiable by knplabs/knp-menu-bundle[v2.1.1].
    - sonata-project/admin-bundle 3.1.0 requires knplabs/knp-menu-bundle ^2.1.1 -> satisfiable by knplabs/knp-menu-bundle[v2.1.1].
    - Conclusion: don't install knplabs/knp-menu-bundle v2.1.1
    - Installation request for sonata-project/admin-bundle ^3.0 -> satisfiable by sonata-project/admin-bundle[3.0.0, 3.1.0].
```

And I expected this to happen:

Having all dependencies installed at lowest version, including `knplabs/knp-menu-bundle` 2.1.1.

If you do a classic `composer update` and then a `composer update --prefer-lowest`, it works. Very weird.

You can see the related Travis job: https://travis-ci.org/sonata-project/SonataDoctrineORMAdminBundle/jobs/130328983#L210

Additional depends info:

``` bash
$ composer depends --tree knplabs/knp-menu-bundle
knplabs/knp-menu-bundle v2.1.1 This bundle provides an integration of the KnpMenu library
└──sonata-project/admin-bundle 3.1.0 (requires knplabs/knp-menu-bundle ^2.1.1)
   └──sonata-project/doctrine-orm-admin-bundle 3.9999999.9999999.9999999-dev (requires sonata-project/admin-bundle ^3.0)
```

**Comment by Soullivaneuh** [comment]
Tested on `1.0.0`, `1.1.0` and `1.1.1`: all failed.

Workaround waiting issue fix: adding `"knplabs/knp-menu-bundle": "^2.1.1"` on the `require` section.

**Comment by Soullivaneuh** [comment]
Got the same issue with sonata-project/block-bundle: https://travis-ci.org/sonata-project/SonataBlockBundle/jobs/131382580#L216

**Comment by Soullivaneuh** [comment]
@Seldaek Just for curiosity, what the `Solver` label means?

**Comment by stof** [comment]
@Soullivaneuh this means that the issue is related to the dependency solver (as opposed to other parts of Composer)

**Comment by Soullivaneuh** [comment]
Same kind of issue with the following `composer.json`:

```json
{
    "name": "sonata-project/seo-bundle",
    "type": "symfony-bundle",
    "description": "Symfony SonataSeoBundle",
    "keywords": ["seo", "sonata"],
    "homepage": "https://sonata-project.org/bundles/seo",
    "license": "MIT",
    "authors": [
        {
            "name": "Thomas Rabaix",
            "email": "thomas.rabaix@sonata-project.org",
            "homepage": "https://sonata-project.org"
        },
        {
            "name": "Sonata Community",
            "homepage": "https://github.com/sonata-project/SonataSeoBundle/contributors"
        }
    ],
    "require": {
        "php": "^5.3 || ^7.0",
        "symfony/framework-bundle": "^2.3 || ^3.0",
        "symfony/options-resolver": "^2.3 || ^3.0",
        "sonata-project/exporter": "^1.2.2",
        "sonata-project/block-bundle": "^2.3.9 || ^3.0",
        "twig/twig": "^1.12"
    },
    "require-dev": {
        "guzzle/guzzle": "3.*",
        "symfony/console": "^2.3 || ^3.0",
        "symfony/finder": "^2.3 || ^3.0",
        "sonata-project/admin-bundle": "^3.0",
        "symfony/phpunit-bridge": "^2.7 || ^3.0",
        "sllh/php-cs-fixer-styleci-bridge": "^2.0"
    },
    "suggest": {
        "guzzle/guzzle": "3.*",
        "knplabs/knp-menu-bundle": "Used by the BreadcrumbMenuBuilder"
    },
    "autoload": {
        "psr-4": { "Sonata\\SeoBundle\\": "" }
    },
    "extra": {
        "branch-alias": {
            "dev-master": "2.x-dev"
        }
    }
}
```

The result:

```bash
$ composer up --prefer-lowest
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - sonata-project/admin-bundle 3.0.0 requires sonata-project/block-bundle ^3.0 -> satisfiable by sonata-project/block-bundle[3.0.0].
    - sonata-project/admin-bundle 3.1.0 requires sonata-project/block-bundle ^3.0 -> satisfiable by sonata-project/block-bundle[3.0.0].
    - Conclusion: don't install sonata-project/block-bundle 3.0.0
    - Installation request for sonata-project/admin-bundle ^3.0 -> satisfiable by sonata-project/admin-bundle[3.0.0, 3.1.0].
```

This is caused by this line: `"sonata-project/block-bundle": "^2.3.9 || ^3.0",`. If you remove `^2.3.9`, it works. But composer should go directly on 3.0 while resolving.

And, as the previous error, this happen only if you don't have `vendor` and `composer.lock` file.

Related Travis job: https://travis-ci.org/sonata-project/SonataSeoBundle/jobs/131432442#L219

**Comment by acrobat** [comment]
Related issue #7161. A workaround is to first do an update and afterwards to the prefer-lowest update

**Comment by naderman** [comment]
Should be fixed with https://github.com/composer/composer/pull/7946 - please verify if you can.

**Comment by acrobat** [comment]
@naderman I can confirm this bug is fixed with the patch provided in #7946. I used this reproducer with `composer update --prefer-dist --prefer-stable --prefer-lowest`

```
{
    "require": {
        "doctrine/lexer": "1.*",
        "egulias/email-validator": "~2.0"
    }
}
```

Thanks, good job! 👌

**Comment by naderman** [comment]
Now fixed in master, available on snapshots, will go into next 1.8 release.
